tensorflow/BUILD

```diff
@@ -485,32 +487,113 @@ tf_cc_shared_object(
 # an "-exported_symbols_list" command.  -z defs disallows undefined
 # symbols in object files.
 
-tf_cc_shared_object(
-    name = "libtensorflow.so",
-    linkopts = select({
-        "//tensorflow:darwin": [
-            "-Wl,-exported_symbols_list",  # This line must be directly followed by the exported_symbols.lds file
-            "$(location //tensorflow/c:exported_symbols.lds)",
-            "-Wl,-install_name,@rpath/libtensorflow.so",
-        ],
-        "//tensorflow:windows": [],
-        "//conditions:default": [
-            "-z defs",
-            "-Wl,--version-script",  #  This line must be directly followed by the version_script.lds file
-            "$(location //tensorflow/c:version_script.lds)",
+SHARED_LIBRARY_NAME_PATTERN = [
+    # prefix, extension
+    ("lib", ".so"),  # On Linux, shared libraries are usually named as libfoo.so
+    ("lib", ".dylib"),  # On macos, shared libraries are usually named as libfoo.dylib
+    ("", ".dll"),  # On Windows, shared libraries are usually named as foo.dll
+]
+
+[
+    tf_cc_shared_object(
```

Comment:
This pattern is repeated in a few places, should we make a separate bazel macro for this (unless `select` can be used for `name` as suggested in a comment below)

Reply:
Good idea! It's done, please take a look.
Basically this pattern approach is a workaround due to cc_binary not able to generate shared library with different names for different platforms. We can remove this workaround when we have a cc_shared_library implemented in Starlark, for that we have to wait for https://github.com/bazelbuild/bazel/issues/4570